pkgs/development/libraries/jemalloc/default.nix

```diff
@@ -49,9 +49,12 @@ stdenv.mkDerivation rec {
 
   env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.isDarwin "-Wno-error=array-bounds";
 
-  # Tries to link test binaries binaries dynamically and fails
+  # static: Tries to link test binaries binaries dynamically and fails
+  # llvm: Tries linking with an undefined symbol and fails.
   doCheck = !stdenv.hostPlatform.isStatic;
 
+  checkFlags = lib.optional (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") "LDFLAGS=-lc++";
+
```

Comment:
so this is not restricted to LLVM17+ as it is an issue for all LLVMs nor does it seem to  be related to lld. the condition doesn't seem correct as it should be something like `libcxx.isLLVM` i think. but i think the underlying issue is that libjemalloc.so is not linking to `libc++.so`.

when built with gcc toolchain, there is a c++ lib linked in but this is not the case with pkgsLLVM

eg:
```console
$ nix-build -A jemalloc
[...]
$ patchelf --print-needed result/lib/libjemalloc.so
libm.so.6
libstdc++.so.6
libgcc_s.so.1
libpthread.so.0
libc.so.6
$ nix-build --expr 'with import ./. {}; pkgsLLVM.jemalloc.overrideAttrs { doCheck = false; }'
[...]
$ patchelf --print-needed result/lib/libjemalloc.so
libm.so.6
libunwind.so.1
libpthread.so.0
libc.so.6
```

so libjemalloc.so when built with pkgsLLVM is not adding `libc++.so`, but one can look at the undefined symbols, `nm -u result/lib/libjemalloc.so` and see that there are undefined symbols that are defined in `libc++.so`. so while checkFlags works, it seems something is off with the build. as noted before, adding `LDFLAGS=-lc++` conditionally to configureFlags would work it is curious why it is not added in the first place.

Reply:
on x64 darwin it seems to link to libc++ fine.
```console
$ nix-build -A jemalloc
[...]
$ otool -L result/lib/libjemalloc.dylib
result/lib/libjemalloc.dylib:
        /nix/store/h4fzlgr53fngqx2dk7j5xwzvc5jc3hdm-jemalloc-5.3.0/lib/libjemalloc.2.dylib (compatibility version 0.0.0, current version 0.0.0)
        /nix/store/2yjr5xx0qn1cd72lc3l1mw8k90fg5lxs-libcxx-16.0.6/lib/libc++.1.0.dylib (compatibility version 1.0.0, current version 1.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
```

Reply:
> so this is not restricted to LLVM17+ as it is an issue for all LLVMs nor does it seem to be related to lld.

Interesting, with `pkgsLLVM` there are a good number of cases where LLVM 17 introduces certain defaults that cause linking with lld to have build issues with various packages. (Notably ncurses)

> so libjemalloc.so when built with pkgsLLVM is not adding `libc++.so`, but one can look at the undefined symbols, `nm -u result/lib/libjemalloc.so` and see that there are undefined symbols that are defined in `libc++.so`. so while checkFlags works, it seems something is off with the build. as noted before, adding `LDFLAGS=-lc++` conditionally to configureFlags would work it is curious why it is not added in the first place.

So add the `LDFLAGS` to both `configureFlags` & `checkFlags`?

Reply:
> the condition doesn't seem correct as it should be something like `libcxx.isLLVM` i think.

Strange, `libcxx.isLLVM` and `pkgsLLVM.libcxx.isLLVM` are true. I think the check was fine before but without the version check.

Reply:
so created jemalloc2 which is just HEAD from https://github.com/jemalloc/jemalloc with pkgsLLVM and it doesn't require any changes and is linking to `libc++`
```console
$ nix-build -A pkgsLLVM.jemalloc2
[...]
$ patchelf --print-needed result/lib/libjemalloc.so
libm.so.6
libunwind.so.1
libc++.so.1
libc++abi.so.1
libpthread.so.0
libc.so.6
```

so i guess either find the change that introduces the fix and patch or add `LDFLAGS` to configureFlags conditionally with a note that it can be removed on update.

with `LDFLAGS` added to `configureFlags` it will not be required for `checkFlags`.



> > the condition doesn't seem correct as it should be something like `libcxx.isLLVM` i think.
> 
> Strange, `libcxx.isLLVM` and `pkgsLLVM.libcxx.isLLVM` are true. I think the check was fine before but without the version check.

ah, ok, then perhaps `stdenv.cc.libcxx.isLLVM or false` is the correct flag. as we don't want to set the flag when using clangStdenv as that doesn't use libcxx.

Reply:
> so i guess either find the change that introduces the fix and patch or add `LDFLAGS` to configureFlags conditionally with a note that it can be removed on update.

This is probably the better and easier solution. I'll add the note to remove it on the next release.

> with `LDFLAGS` added to `configureFlags` it will not be required for `checkFlags`.

Alright, I'll remove the `checkFlags`.

Reply:
this looks like it could be the PR that does the fix: https://github.com/jemalloc/jemalloc/pull/2348
https://github.com/jemalloc/jemalloc/commit/4422f88d17404944a312825a1aec96cd9dc6c165

Reply:
Oh then we could just fetch the patch and add it in.

Reply:
yeah, just confirmed that patch works.

Reply:
Amazing, I added the patch.